presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java

```diff
@@ -366,15 +371,36 @@ public String toString()
 
     private void ensureWritableBytes(int minWritableBytes)
     {
-        if (bufferPosition + minWritableBytes > slice.length()) {
+        int neededBufferSize = bufferPosition + minWritableBytes;
+        if (neededBufferSize <= slice.length()) {
+            return;
+        }
+
+        if (slice.length() >= maxBufferSize) {
             flushBufferToOutputStream();
+            return;
+        }
+
+        // grow the buffer size
+        int newBufferSize = min(max(slice.length() * 2, minWritableBytes), maxBufferSize);
+        if (neededBufferSize <= newBufferSize) {
+            // we have capacity in the new buffer; just copy the data to the new buffer
+            byte[] previousBuffer = buffer;
+            buffer = new byte[newBufferSize];
+            System.arraycopy(previousBuffer, 0, buffer, 0, bufferPosition);
         }
+        else {
+            // there is no enough capacity in the new buffer; flush the data and allocate the new buffer
+            flushBufferToOutputStream();
+            buffer = new byte[newBufferSize];
+        }
+        slice = wrappedBuffer(buffer);
```

Comment:
I would place this directly below the assignment of `buffer` to make it clear that they are always linked